drivers/net/w5500.c

```diff
@@ -0,0 +1,2232 @@
+#define SN_SR_SOCK_MACRAW      0x42
+
+#define SN_SR_SOCK_SYNSENT     0x15    /* Transitional status */
+#define SN_SR_SOCK_SYNRECV     0x16
+#define SN_SR_SOCK_FIN_WAIT    0x18
+#define SN_SR_SOCK_CLOSING     0x1a
+#define SN_SR_SOCK_TIME_WAIT   0x1b
+#define SN_SR_SOCK_LAST_ACK    0x1d
+
+/* Socket n RX Buffer Size Register (SN_RXBUF) */
+
+#define SN_RXBUF_0KB          0
+#define SN_RXBUF_1KB          1
+#define SN_RXBUF_2KB          2
+#define SN_RXBUF_4KB          4
+#define SN_RXBUF_8KB          5
+#define SN_RXBUF_16KB         16
+
+/* Socket n TX Buffer Size Register (SN_TXBUF) */
+
+#define SN_TXBUF_0KB          0
+#define SN_TXBUF_1KB          1
+#define SN_TXBUF_2KB          2
+#define SN_TXBUF_4KB          4
+#define SN_TXBUF_8KB          5
+#define SN_TXBUF_16KB         16
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* The w5500_driver_s encapsulates all state information for a single
+ * hardware interface
+ */
+
+struct w5500_driver_s
+{
+  bool sk_bifup;               /* true:ifup false:ifdown */
+  struct wdog_s sk_txtimeout;  /* TX timeout timer */
+  struct work_s sk_irqwork;    /* For deferring interrupt work to the work queue */
+  struct work_s sk_pollwork;   /* For deferring poll work to the work queue */
+
+  /* Ethernet frame transmission buffer management */
+
+  uint16_t txbuf_offset[NUM_TXBUFS + 1];
+  uint8_t  txbuf_rdptr;
+  uint8_t  txbuf_wrptr;
+
+  /* The hardware interconnect to the W5500 chip */
+
+  FAR struct spi_dev_s           *spi_dev; /* SPI hardware access    */
+  FAR const struct w5500_lower_s *lower;   /* Low-level MCU specific */
+
+  /* This holds the information visible to the NuttX network */
+
+  struct net_driver_s sk_dev;  /* Interface understood by the network */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* These statically allocated structures would mean that only a single
+ * instance of the device could be supported.  In order to support multiple
+ * devices instances, this data would have to be allocated dynamically.
+ */
+
+/* A single packet buffer per device is used in this example.  There might
+ * be multiple packet buffers in a more complex, pipelined design.  Many
+ * contemporary Ethernet interfaces, for example,  use multiple, linked DMA
+ * descriptors in rings to implement such a pipeline.  This example assumes
+ * much simpler hardware that simply handles one packet at a time.
+ *
+ * NOTE that if CONFIG_NET_W5500_NINTERFACES were greater than 1, you would
+ * need a minimum on one packet buffer per instance.  Much better to be
+ * allocated dynamically in cases where more than one are needed.
+ */
+
+static uint8_t g_pktbuf[MAX_NETDEV_PKTSIZE + CONFIG_NET_GUARDSIZE];
```

Comment:
I think it is good to either
```
static uint16_t g_pktbuf[(MAX_NETDEV_PKTSIZE + CONFIG_NET_GUARDSIZE + 1) / 2];
```
or
```
static uint8_t g_pktbuf[MAX_NETDEV_PKTSIZE + CONFIG_NET_GUARDSIZE] aligned_data(2);
```
since net stack assumes that data are half word aligned. Please see `struct cdcecm_driver_s` for similar issue.
I will update skeleton code.